source/agora/consensus/validation/Block.d

```diff
@@ -113,6 +119,282 @@ public string isInvalidReason (const ref Block block, in ulong prev_height,
+        - Block height is 0
+        - The previous block hash of the block must be empty
+        - The number of transactions in the block is above zero and less than
+          or equal to Block.TxsInBlock
+        - All the transactions pass validation, which implies:
+            - Transaction type in block should only be Payment or Freeze
+            - The inputs of the transaction(s) are empty
+            - There is one or more outputs of the transaction
+        - All the enrollments pass validation, which implies:
+            - The enrollments refer to freeze tx's in this block
+
+    Params:
+        block = The genesis block to check
+
+    Returns:
+        `null` if the genesis block is valid, otherwise a string explaining
+        the reason it is invalid.
+
+*******************************************************************************/
+
+public string isValidGenesisBlock (immutable Block* block) nothrow @safe
+{
+    if (block.header.height != 0)
+        return "GenesisBlock: The height of the block is not 0";
+
+    if (block.header.prev_block != Hash.init)
+        return "GenesisBlock: Header.prev_block is not empty";
+
+    if (block.txs.length == 0)
+        return "GenesisBlock: Transaction(s) are empty";
+
+    if (!block.txs.isSorted())
+        return "GenesisBlock: Transactions are not sorted";
+
+    if (block.txs.length > Block.TxsInBlock)
+        return "GenesisBlock: The number of transactions is out of bounds";
+
+    UTXOSetValue[Hash] utxo_set;
+    foreach (const ref tx; block.txs)
+    {
+        if (!(tx.type == TxType.Payment || tx.type == TxType.Freeze))
+            return "GenesisBlock: Transaction type allows only "
+                ~ "Payment and Freeze";
+
+        if (tx.inputs.length != 0)
+             return "GenesisBlock: Not empty input(s) in the transaction";
+
+        if (tx.outputs.length == 0)
+            return "GenesisBlock: No output(s) in the transaction";
+
+        Hash tx_hash = tx.hashFull();
+        foreach (idx, const ref output; tx.outputs)
+        {
+            // disallow negative amounts
+            if (!output.value.isValid())
+                return "GenesisBlock: Output(s) overflow or underflow"
+                    ~ "in the transaction";
+
+            // disallow 0 amount
+            if (output.value == Amount(0))
+                return "GenesisBlock: Value of output is 0"
+                    ~ "in the transaction";
+
+            const UTXOSetValue utxo_value = {
+                unlock_height: 0,
+                type: tx.type,
+                output: output
+            };
+            utxo_set[UTXOSetValue.getHash(tx_hash, idx)] = utxo_value;
+        }
+    }
+
+    Hash[] merkle_tree;
+    if (block.header.merkle_root !=
+        Block.buildMerkleTree(block.txs, merkle_tree))
+        return "GenesisBlock: Merkle root does not match header's";
+
+    // If there are no enrollments, return them here early.
```

Comment:
This conflicts with https://github.com/bpfkorea/agora/pull/875. Depending on which PR gets merged first, this code will have to be fixed somewhere.